akka-remote/src/main/scala/akka/remote/artery/SendQueue.scala

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.util.Queue
+import akka.stream.stage.GraphStage
+import akka.stream.stage.OutHandler
+import akka.stream.Attributes
+import akka.stream.Outlet
+import akka.stream.SourceShape
+import akka.stream.stage.GraphStageLogic
+import org.agrona.concurrent.ManyToOneConcurrentArrayQueue
+import akka.stream.stage.GraphStageWithMaterializedValue
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueueTail
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+import scala.annotation.tailrec
+
+/**
+ * INTERNAL API
+ */
+private[remote] object SendQueue {
+  trait QueueValue[T] {
+    def inject(queue: Queue[T]): Unit
+    def offer(message: T): Boolean
+  }
+
+  private trait WakeupSignal {
+    def wakeup(): Unit
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class SendQueue[T] extends GraphStageWithMaterializedValue[SourceShape[T], SendQueue.QueueValue[T]] {
+  import SendQueue._
+
+  val out: Outlet[T] = Outlet("SendQueue.out")
+  override val shape: SourceShape[T] = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, QueueValue[T]) = {
+    @volatile var needWakeup = false
+    @volatile var queue: Queue[T] = null
```

Comment:
It's set from the outside, in the `inject` method. See https://github.com/akka/akka/pull/20657/files/07dc5725a96cd7809f72cf74de0b6b19ac6de015#diff-8a8f558f6dd31f0cc55c0cf89a3c9446R94

The reason is that otherwise we have the problem with that callbacks can't be invoked before the stage is fully started.